### DISCUSSION TITLE
how to update the position of fireeffect without making it regenerates

### DISCUSSION THREAD

**Discussion by XianranZ** [openscenegraph/OpenSceneGraph#1340] [Q&A]
Hi, everyone. I am using osg3.4 and osgearth2.7 to develop a digital 3D earth.I want to make a FireEffect(osgParticle::FireEffect) on the earth surface, and first problem occurred. when I used matrixtransform to transform the position of the FireEffect, it didn't work. so I had to use the FireEffect::setposition to set it's position. I want to know the matrixtransform didn't work? Does anyone know that?
After that I want to update the position of the FireEffect. But when I change the FireEffect's position, it will be regenerated. Does anyone know how I can do?
thanks a lot!

**Comment by xarray** [comment]
In order to make the particle system moving together with the MatrixTransform, you have to place the ModularEmitter as the transform's child, not the effect itself.
The osgParticle library is also outdated as it is designed about 20 years ago. Maybe you could consider some modern solutions to integrated with OSG/VSG, like Effekseer.

**Comment by XianranZ** [comment]
> In order to make the particle system moving together with the MatrixTransform, you have to place the ModularEmitter as the transform's child, not the effect itself.
> The osgParticle library is also outdated as it is designed about 20 years ago. Maybe you could consider some modern solutions to integrated with OSG/VSG, like Effekseer.

Thank you so much for your reply!

I put the FireEffect's ModularEmitter into the MT, but unfortunately it doesn't work.
Here are the steps I tried!

1. add the emitter into the MT
```
mt->addChild(rpFireEffect->getEmitter());
mt->addChild(rpFireEffect.get());
```
it doesn't work! The position has no change.

2. The return type of getEmitter is osgParticle::Emitter, which is the parent of osgParticle::ModularEmitter. So I make a cast.
```
osgParticle::ModularEmitter* pme = dynamic_cast<osgParticle::ModularEmitter*>(rpFireEffect->getEmitter());
if(pme != NULL)
{
    mt->addChild(pme.get());
    mt->addChild(rpFireEffect.get());
}
```
it doesn't work either.

I've been stuck on this for days, so I did some research and read the source code of OSG, and found some interesting phenomena.
1. There are two types transform in OSG, which are osg::MatrixTransform and osg::PositionAttitudeTransform, respectively. The former is a relative transformation, the latter is an absolute transformation. A relative transformation means that it only transforms relative to its parent node, while an absolute transformation means that it transforms in the osg world coordinate system, and this transformation is not affected by the parent node. I wrote the following code to validate my point.
```
osg::ref_ptr<osg::PositionAttitudeTransform> pat = osg::PositionAttitudeTransform;
pat->setPosition(osg::Vec3d(10.0, 0.0, 0.0));
pat->addChild(rpFireEffect.get());
root->addChild(pat.get());
root->addChild(rpFireEffect.get());
```
Here is the result.

**Reply by xarray** [reply]
You may have to put the emitter under a transform, and the effect node under the root. I personally don't use these classes so I could only provide some old and simple code segments for validation, which surely works:

> osg::Group* createParticles(osgParticle::ParticleSystem* ps)
> {
>     osg::StateSet* stateset = ps->getOrCreateStateSet();
>     stateset->setAttribute(new osg::Point(20.0f));
>     stateset->setTextureAttributeAndModes(0, new osg::PointSprite, osg::StateAttribute::ON);
>     ps->getDefaultParticleTemplate().setLifeTime(5.0f);
>     ps->getDefaultParticleTemplate().setShape(osgParticle::Particle::POINT);
>     ps->setDefaultAttributesUsingShaders("Images/smoke.rgb", true, 0);
> 
>     //  ......  omitted
> 
>     osg::ref_ptr<osgParticle::ModularEmitter> emitter = new osgParticle::ModularEmitter;
>     emitter->setParticleSystem(ps);
>     emitter->setCounter(rrc.get());
>     emitter->setShooter(shooter.get());
> 
>     osg::ref_ptr<osgParticle::ModularProgram> program = new osgParticle::ModularProgram;
>     program->setParticleSystem(ps);
>     program->addOperator(accel.get());
> 
>     osg::ref_ptr<osg::MatrixTransform> parent = new osg::MatrixTransform;
>     parent->addChild(emitter.get());
>     parent->addChild(program.get());
>     return parent.release();
> }

You will find that ParticleSystemUpdater and ParticleSystem are not here. They are added to the root node. The ParticleEffect class includes these components inside so you may have to manually manage them or write your own implementations for specific uses.

Also I don't think your descriptions about MatrixTransform and PositionAttitudeTransform are correct. They can be either relative or absolute transformations, with the setReferenceFrame() method.

**Reply by XianranZ** [reply]
> You may have to put the emitter under a transform, and the effect node under the root. I personally don't use these classes so I could only provide some old and simple code segments for validation, which surely works:
> 
> > osg::Group* createParticles(osgParticle::ParticleSystem* ps)
> > {
> > osg::StateSet* stateset = ps->getOrCreateStateSet();
> > stateset->setAttribute(new osg::Point(20.0f));
> > stateset->setTextureAttributeAndModes(0, new osg::PointSprite, osg::StateAttribute::ON);
> > ps->getDefaultParticleTemplate().setLifeTime(5.0f);
> > ps->getDefaultParticleTemplate().setShape(osgParticle::Particle::POINT);
> > ps->setDefaultAttributesUsingShaders("Images/smoke.rgb", true, 0);
> > ```
> > //  ......  omitted
> > 
> > osg::ref_ptr<osgParticle::ModularEmitter> emitter = new osgParticle::ModularEmitter;
> > emitter->setParticleSystem(ps);
> > emitter->setCounter(rrc.get());
> > emitter->setShooter(shooter.get());
> > 
> > osg::ref_ptr<osgParticle::ModularProgram> program = new osgParticle::ModularProgram;
> > program->setParticleSystem(ps);
> > program->addOperator(accel.get());
> > 
> > osg::ref_ptr<osg::MatrixTransform> parent = new osg::MatrixTransform;
> > parent->addChild(emitter.get());
> > parent->addChild(program.get());
> > return parent.release();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > }
> 
> You will find that ParticleSystemUpdater and ParticleSystem are not here. They are added to the root node. The ParticleEffect class includes these components inside so you may have to manually manage them or write your own implementations for specific uses.
> 
> Also I don't think your descriptions about MatrixTransform and PositionAttitudeTransform are correct. They can be either relative or absolute transformations, with the setReferenceFrame() method.

Thanks a lots! And I think your answer is correct! I use the emitter and particle system to custom a new fire effect, and it works! what's more, I find a new way to change the position and posture of osgParticle::FireEffect, which is to use its own method named setPosition(), it works!
And I make some tests to check the differences between MatrixTransform and PositionAttitudeTransform. Actually, they have same functions with  the setReferenceFrame() method.
Thanks a again!
